[Day3]

terry-yes opened the pull request:
## Day3

- 셀에 표시될 정보가 기존의 "이미 갖고 있던 JSON파일"에서 "통신해서 가져온 JSON파일"로 변경되었습니다.
- 셀을 클릭했을 때 표시될 상세 화면의 뷰를 만들고 있습니다.
![](https://images.velog.io/images/taelee/post/08be029e-2bcd-49d0-b57b-0831c0ccf608/Feb-07-2022%2017-58-00.gif)

terry-yes committed:
feat: 기존 JSON파일에서 셀정보를 가져오던 것을 통신한 JSON파일에서 가져오는 것으로 변경

terry-yes committed:
feat: foodListVC을 Navigation controller에 embed함

terry-yes committed:
feat: collectionView의 cell을 클릭했을 때 다음화면으로 이동하도록 구현

terry-yes committed:
feat: Detail뷰에서 hashID를 통한 JSON 통신 구현

terry-yes committed:
feat: detail화면 뷰 만드는중

terry-yes committed:
docs: day3 구현사항 README.md에 정리

terry-yes committed:
feat: detail화면 뷰 구현

- 아직 하단의 사진들은 제대로 구현되지 않음
- 이미지캐싱을 완성하고 구현할 예정

terry-yes committed:
feat: 셀의 음식 이미지를 직접만든 cache에서 NSCache를 사용하는 것으로 변경

terry-yes committed:
feat: detail화면에서 썸네일 이미지가 보이도록 구현

terry-yes committed:
feat: cacheManager를 만들어서 이미지를 캐싱하도록 구현

bamiboo-han commented on Onban/Onban/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,132 @@
+                    self.titleLabel.text = data.data.productDescription
+                    self.detailLabel.text = data.data.productDescription
+                    self.originalPrice.text = data.data.prices[0]
+                    self.currentPrice.text = data.data.prices[1]
```
인덱스 1이 없는 경우가 있어 오류가 발생하고 있습니다~

terry-yes replied:
수정했습니다!😁

bamiboo-han commented on Onban/Onban/ViewModels/FoodViewModel.swift:
```diff
@@ -10,6 +10,7 @@ import UIKit
+    var hashID: String { food.hashID }
     var title: String { food.title }
     var description: String { food.description}
     var nPrice: String? {
```
이번 PR 에는 반영되지 않는 내용이라 여기에 comment 달아둡니다~
nPrice, sPrice 타입이 String 으로 "원" 단위도 같이 넘어오고 있어서 "원"이 두번 노출되네요! ㅎㅎ
또한 isNew 판단 값이 "론칭특가" 로 들어가 있어서 "런칭특가" 태그가 미노출되고 있습니다~

terry-yes replied:
수정했습니다!😁

bamiboo-han commented:
딱히 개발적으로 리뷰달건 찾지 못해서 ㅎㅎ 추가로 알고계시면 좋을것 같아 댓글 달아둡니다~ 

메모리 관리 관련된 내용인데요 클로저 사용시 [weak self] 에 대해 찾아보시면 좋을것 같아요!
클로저 내부에서 referece count 가 증가해 deinit 되지 않는 이슈가 있을 수 있습니다

terry-yes committed:
feat: 컬렉션 뷰의 음식 썸네일 이미지를 앱이 실행되자마자 전체 다운받도록 변경

terry-yes committed:
feat: 컬렉션뷰의 네비게이션 바가 표시되지 않도록 변경

terry-yes committed:
feat: detail뷰의 이미지도 캐싱 처리

terry-yes committed:
feat: detail뷰에서는 네비게이션바가 보이도록 구현

terry-yes committed:
feat: detail뷰에 하단 이미지 높이를 원본 이미지 비율대로 표시되도록 구현

ashe-0 commented on Onban/Onban/Controllers/FoodListViewController.swift:
```diff
@@ -34,35 +34,18 @@ class FoodListViewController: UIViewController {
+        self.foodListVM = FoodListViewModel(foodsList: [])
+        foodListVM?.addFoodViewModel(type: .main) {
+            self.foodListVM?.addFoodViewModel(type: .soup, completion: {
+                self.foodListVM?.addFoodViewModel(type: .side, completion: {
```
순서가 중요하지 않은 비동기 작업 여러 개를 함께 처리하기 위해서는 dispatchgroup에 관한 내용을 참고해봐도 좋을 것 같습니다. ㅎ-ㅎ

terry-yes replied:
감사합니다!!
dispatchGroup을 사용해봤는데 이렇게 하는건지 확신이 안드네요
좀 더 공부해보면서 수정해나가도록 하겠습니다.

ashe-0 commented on Onban/Onban/ViewModels/FoodViewModel.swift:
```diff
@@ -45,10 +45,32 @@ class FoodViewModel {
     case side
     case soup
+    
+    func JSONURL() -> String {
```
단순한 string값 반환의 경우 function 대신 computed property 형태로 표현하면 더 직관적일 듯 합니다 :)  
( headerTitle()의 경우도 같습니다 )

terry-yes replied:
수정했습니다!😁

terry-yes committed:
fix: detail화면에서 사진갯수에 따라 크래시가 나던 문제 해결

terry-yes committed:
fix: detail 화면에서 음식 가격이 한개만 제공될 때 발생되던 에러 해결

terry-yes committed:
feat: detail 화면에서 이전화면의 title정보를 넘겨받아 출력될 수 있도록 구현

terry-yes committed:
feat: detail 화면에서 이벤트특가 뱃지와 런칭특가 뱃지가 뱃지유무와 상관없이 표현되던것에서 제대로 표현되도록 변경

terry-yes committed:
feat: 초기 음식 데이터를 가져오는 부분에 콜백을 여러번 중첩해서 쓰던 것에서 DispathGroup을 활용

terry-yes committed:
refactor: FoodViewModel에서 인자없는 함수들을 Computed Property로 바꿈

terry-yes commented:
> 딱히 개발적으로 리뷰달건 찾지 못해서 ㅎㅎ 추가로 알고계시면 좋을것 같아 댓글 달아둡니다~
> 
> 메모리 관리 관련된 내용인데요 클로저 사용시 [weak self] 에 대해 찾아보시면 좋을것 같아요! 클로저 내부에서 referece count 가 증가해 deinit 되지 않는 이슈가 있을 수 있습니다

리뷰 감사합니다! self를 달 때 weak를 쓸지 말지 생각을 안 했었는데
고민해서 달도록 하겠습니다!
아직 수정하지는 않았지만 공부하면서 수정해나가도록 하겠습니다~!

terry-yes committed:
chore: 쓰지 않게 된 코드 삭제

terry-yes committed:
feat: detail 뷰에서 맨 위 thumnail이미지의 가로 스크롤이 가능하도록 구현

terry-yes committed:
feat: detail 화면 썸네일 이미지에 이미지가 페이지단위로 넘어가게 구현

terry-yes committed:
feat: 이미지 캐싱시 압축하여 저장하도록 변경

ashe-0 commented on Onban/Onban/Controllers/FoodListViewController.swift:
```diff
@@ -47,15 +47,32 @@ class FoodListViewController: UIViewController {
+        foodListVM?.addFoodViewModel(type: .side) {
+            DispatchQueue.main.async {
+                self.collectionView.reloadData()
+            }
```
모든 작업이 완료된 후 최종적으로 수행할 작업은 group.notify 에서 작성하시면 됩니다!  짧은 시간 안에 반영하셨네요 👍 👍

terry-yes replied:
감사합니다!! 말씀해주신거 적용해서 바꾸었습니다!!
이 부분이 고민이 많았던 부분인데 알려주셔서 감사해요~

terry-yes committed:
feat: detail 화면의 thumnail이미지가 자동으로 스크롤 되도록 구현

- 문제: 두번째 썸네일 이미지로 스크롤 됐을 때 이미지가 화면의 중앙에 있지 않은 문제 고쳐야함

kyoo-sl commented on Onban/Onban/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,407 @@
+    let currentPriceLabel: UILabel = {
+        let label = UILabel()
+        label.font = UIFont.boldSystemFont(ofSize: 28)
+        return label
```
currentPriceLabel에 gray설정이 빠져서 
다크모드에서는 안 보이는 것 같아요~

terry-yes replied:
수정하였습니다!!😁

kyoo-sl commented on Onban/Onban/Extension/UIImage+AspectRatio.swift:
```diff
@@ -0,0 +1,16 @@
+    var aspectRatio: CGFloat {
+        let width = self.size.width
+        let height = self.size.height
+        return width / height
```
나누는 수는 0이 아닌지 체크하면 더 안정적일 것 같아요~

terry-yes replied:
수정하였습니다!!😁
리턴형을 옵셔널로 바꾸어서 0으로 나누려고 시도하면 nil을 리턴하도록 했습니다!

kyoo-sl commented on Onban/Onban/Models/FoodDescription.swift:
```diff
@@ -0,0 +1,36 @@
+        case deliveryInfo = "delivery_info"
+        case deliveryFee = "delivery_fee"
+        case prices
+        case detailSection = "detail_section"
```
규칙이 동일하다면, 여기서 굳이 string을 지정하지 않고, JSONDecoder에서 아래처럼 설정하는 방법도 있어요~ `JSONDecoder().keyDecodingStrategy = .convertFromSnakeCase`

terry-yes replied:
감사합니다!! 해당 부분 말씀해주신대로 수정하였습니다!
훨씬 더 편하게 Decodable 모델을 만들 수 있게 되었어요!

kyoo-sl reviewed (approved):
확인했습니다, 수고하셨어요 테리👍🏻

terry-yes committed:
fix: DispatchGroup에 group.notify()를 추가하여 모든 비동기 작업이 끝났을 때 일어날 코드 구현

terry-yes committed:
feat: 다크모드에 대응할 수 있도록 색상을 system색상으로 변경

terry-yes committed:
refactor: 코드 리뷰 사항 반영

- decoder사용시 keyDecodingStrategy를 사용하여 단순하 snakeCase Decodable모델을 처리
- 0으로 나눌려고 시도할 때 에러가 나지 않게 옵셔널 처리

terry-yes commented:
> 확인했습니다, 수고하셨어요 테리👍🏻

꼼꼼한 리뷰 감사드려요!! 
4주동안 리뷰해주셔서 감사드립니다!! 😁
